### PR TITLE
Kazuda Xiono, Best Pilot in the Galaxy

### PR DESCRIPTION
**Pull request by moyerr**
## Summary of changes

 - Implement a mechanism for "lose all abilities"
 - Implement a mechanism that allows a player to take an extra action
 - Implement Kazuda Xiodo leader, utilizing the above
 - Add a comprehensive test suite for scenarios where a unit "loses all abilities"
 - Add a `CardRoundLastingEffectSystem` (No Good to Me Dead is updated to use this)

Note: One test case is marked as `pending` because blanking keywords needs a little extra work

| Undeployed | Deployed |
| :---: | :---: |
|  |  |

### CODE REVIEW

File: server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                AbilityHelper.immediateEffects.selectCard({
+                    cardTypeFilter: WildcardCardType.Unit,
+                    controller: RelativePlayer.Self,
+                    innerSystem: AbilityHelper.immediateEffects.forThisRoundCardEffect({
+                        effect: AbilityHelper.ongoingEffects.loseAllAbilities()
+                    })
+                }),
+                AbilityHelper.immediateEffects.forThisPhaseCardEffect({
+                    effect: AbilityHelper.ongoingEffects.additionalAction()
+                })
```
Comment (moyerr) [inline server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts:28]: I set this to a `forThisPhaseCardEffect` somewhat arbitrarily. It really just needs to be a "long enough to outlast this turn" effect, but we don't have anything shorter than phase AFAIK.

File: server/game/core/gameSteps/phases/ActionPhase.ts [hunk 2]
```diff
@@ -35,7 +35,15 @@ export class ActionPhase extends Phase {
...
         // breaks the action loop if both players have passed
-        this.game.queueSimpleStep(() => this.game.rotateActivePlayer(), 'rotateActivePlayer');
+        this.game.queueSimpleStep(() => {
+            const activePlayer = this.game.getActivePlayer();
+            if (activePlayer && activePlayer.hasOngoingEffect(EffectName.AdditionalAction)) {
+                activePlayer.removeOngoingEffects(EffectName.AdditionalAction);
+            } else {
+                this.game.rotateActivePlayer();
+            }
+        }, 'rotateActivePlayer');
```
Comment (moyerr) [inline server/game/core/gameSteps/phases/ActionPhase.ts:45]: This is a naive implementation that only supports a single additional action. The `additionalAction` effect already supports an `amount`, so we could reduce by 1 here instead, but we don't have a case for that yet.

File: test/scenarios/ability/LoseAllAbilities.spec.ts [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+                expect(context.player1).toBeAbleToSelectExactly([
+                    context.p2Base,
+                    context.consularSecurityForce,
+                    context.academyDefenseWalker
+                ]);
+                context.player1.clickCard(context.consularSecurityForce);
+            });
+        });
+
+        describe('Some meta-relavant and edge cases worth covering:', function() {
```
Comment (AMMayberry1) [inline test/scenarios/ability/LoseAllAbilities.spec.ts:515]: For this kind of thing we tend to put them in the `scenarios/` folder, maybe under a new test file

File: test/scenarios/ability/LoseAllAbilities.spec.ts [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+                            'trench-run',
+                            'traitorous'
+                        ],
+                        groundArena: [
+                            // 'academy-defense-walker',
+                            // 'contracted-hunter',
+                            // 'grogu#irresistible',
+                            // 'liberated-slaves',
+                            // '97th-legion#keeping-the-peace-on-sullust',
+                            // 'battlefield-marine'
```
Comment (AMMayberry1) [inline test/scenarios/ability/LoseAllAbilities.spec.ts:533]: These needed?

File: test/server/cards/04_JTL/leaders/KazudaZionoBestPilotInTheGalaxy.spec.ts [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+                context.player1.clickPrompt('Choose any number of friendly units');
+                expect(context.player1).toHavePrompt('Choose friendly units that will lose all abilities for this round');
+
+                // Remove his own abilities (among others)
+                context.player1.clickCard(context.kazudaXiono);
+                context.player1.clickCard(context.contractedHunter);
+                context.player1.clickCard(context.fireball);
+                context.player1.clickPrompt('Done');
+
+                // No damage was restored because Kazuda lost his abilities
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/leaders/KazudaZionoBestPilotInTheGalaxy.spec.ts:197]: Will follow up on this in chat but I think this actually wrong due to this ruling:

image

File: test/server/cards/04_JTL/leaders/KazudaZionoBestPilotInTheGalaxy.spec.ts [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+                context.player1.clickCard(context.contractedHunter);
+                context.player1.clickCard(context.fireball);
+                context.player1.clickPrompt('Done');
+
+                // No damage was restored because Fireball lost its abilities
+                expect(context.p1Base.damage).toBe(10);
+
+                context.player2.passAction();
+
+                // Use Frontline Shuttle's ability to attack with Fireball again
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/leaders/KazudaZionoBestPilotInTheGalaxy.spec.ts:387]: Just a note for the future, we're typically okay with setting `exhausted = false` in tests (it's one of the few explicit game state changes I feel comfortable with)

File: server/game/core/card/NonLeaderUnitCard.ts [hunk 7]
```diff
@@ -73,8 +77,23 @@ export class NonLeaderUnitCard extends NonLeaderUnitCardParent implements INonLe
...
+        return this.isBlank() ? [] : super.getActionAbilities();
+    }
+
+    public override getTriggeredAbilities(): TriggeredAbility[] {
+        return this.isBlank() ? [] : super.getTriggeredAbilities();
+    }
+
+    public override getConstantAbilities(): IConstantAbility[] {
+        return this.isBlank() ? [] : super.getConstantAbilities();
+    }
```
[the remark this answers is not in the document]
Comment (moyerr) [reply server/game/core/card/NonLeaderUnitCard.ts:98]: Oh of course. I'll make this change!

File: server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts (hunk 1 above)
Comment (moyerr) [reply server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts:28]: @AMMayberry1 Huh... `Duration.Persistent` doesn't seem to work, but changing it to a `playerLastingEffect()` with `Duration.UntilEndOfPhase` does work

File: test/server/cards/04_JTL/leaders/KazudaZionoBestPilotInTheGalaxy.spec.ts (hunk 5 above)
Comment (moyerr) [reply test/server/cards/04_JTL/leaders/KazudaZionoBestPilotInTheGalaxy.spec.ts:197]: Issue created: https://github.com/SWU-Karabast/forceteki/issues/805

File: server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts (hunk 1 above)
Comment (AMMayberry1) [reply server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts]: Hmm interesting, there are a couple reasons that could be the case. The ongoing effect system is kinda half-baked overall tbh.

Could you leave a TODO comment for investigating that? It doesn't make a huge difference at the end of the day, just a cleanliness thing

Comment (moyerr) [reply server/game/cards/04_JTL/leaders/KazudaXionoBestPilotInTheGalaxy.ts]: Done!

**Review by AMMayberry1 — APPROVED** [review]
